src/syntax/Parser.js

```diff
@@ -313,6 +315,43 @@ function containsInitializer(declarations) {
   return declarations.some((v) => v.initializer);
 }
 
+const FUNCTION_STATE_SCRIPT = 1;
+const FUNCTION_STATE_MODULE = 1 << 1;
+const FUNCTION_STATE_FUNCTION = 1 << 2;
+const FUNCTION_STATE_ARROW = 1 << 3;
+const FUNCTION_STATE_METHOD = 1 << 4;
+const FUNCTION_STATE_DERIVED_CONSTRUCTOR = 1 << 5;
+const FUNCTION_STATE_GENERATOR = 1 << 6;
+const FUNCTION_STATE_ASYNC = 1 << 7;
+
+/**
+ * This is used to track the functions as the parser descends. It allows
+ * us to determine if we are in a function and what kind of function it is.
+ * This is used to determine if `return` and `super` are allowed.
```

Comment:
Maybe:
A marker type for functions that have special rules for `return` and `super`; set as the parser descends.